feat: Knots support

kwsantiago opened the pull request:
Closes #1302

### Description

This PR adds full support for Bitcoin Knots v29 as a Bitcoin backend implementation in Polar, alongside the existing Bitcoin Core support.

**What changed:**
- Added Bitcoin Knots (v29, v28, v27, v26) as a new managed node implementation
- Created Docker configuration and build files for bitcoind-knots
- Added Bitcoin Knots input field to the network creation form UI
- Updated all type definitions and test files to support `bitcoindKnotsNodes` parameter
- Bitcoin Knots nodes use the same Bitcoin RPC service as Bitcoin Core (they're API-compatible)

**Technical details:**
- New implementation type: `'bitcoind-knots'` added to `BitcoinNode` interface
- Created `createBitcoindKnotsNetworkNode()` function for network node creation
- Updated network store to handle bitcoindKnotsNodes in network creation flow
- Updated test files to include the new parameter
- Port configuration matches bitcoind (RPC, P2P, ZMQ Block, ZMQ Tx)

### Steps to Test

1. Start Polar with `yarn dev`
2. Click "Create Lightning Network"
3. In the form, set Bitcoin Core to 0 and Bitcoin Knots to 1
4. Add 2 LND nodes
5. Click "Create Network"
6. Start the network
7. Verify the backend node is using implementation "bitcoind-knots"
8. Test mining blocks and sending funds - should work identically to Bitcoin Core

### Screenshots
<img width="1882" height="741" alt="image" src="https://github.com/user-attachments/assets/f7bc76ee-b4e5-4a7e-ba21-0a27c85fb9d5" />
<img width="1380" height="277" alt="image" src="https://github.com/user-attachments/assets/f2fa46e4-56b6-4bfc-910b-43283a3a1a6b" />
<img width="1489" height="523" alt="image" src="https://github.com/user-attachments/assets/836c1793-edea-46eb-a2bc-bfdd85c4b798" />

jamaljsr commented:
@kwsantiago Thanks for the contribution. I don't have any objections to adding Bitcoin Knots since the RPC is identical to Bitcoin Core, so there should be very little code changes required for it to work. But I do have some concerns about the ongoing maintenance burden this will add for maintainers (primarily myself) to ensure we add support for future versions of Knots in a timely manner. I'm already behind on this with the 6 implementations we currently support. Would you be willing to commit to helping in this regard for some time?

I've tried to review this but I'm facing a few hurdles getting this working:
1. The `Dockerfile` doesn't build as-is using the latest version `29.0` listed in the README. There is no downloadable file for `29.0` on https://bitcoinknots.org/files/29.x/. I managed to get the image to build by using `29.2` but this breaks other things below.
2. After updating the `defaultRepoState` in [constants.ts](https://github.com/kwsantiago/polar/blob/pr/kwsantiago/1303/src/utils/constants.ts#L428-L429) to use `29.2`, I was getting an error creating a network because the [compatibility](https://github.com/kwsantiago/polar/blob/pr/kwsantiago/1303/src/utils/constants.ts#L399-L403) mapping for LND targets `bitcoind` versions. Since there is no `29.0` version of Knots, I had to change this to `29.2`. Unfortunately, this breaks using Core nodes since we haven't added support for `29.2` yet. I suspect this will be a challenge going forward as we'll have to keep the versions of both in sync at all times for the compatibility check to work, or we'll have to do a refactoring of how compatibility is defined and validated to handle differing Bitcoin node versions. 
3. After fixing the above, I was able to create the network. When i tried to start it, I get the docker error below. It looks like the Knots nodes are incorrectly referencing the `bitcoind` image instead of `bitcoind-knots`.
    ```
    Error response from daemon: manifest for polarlightning/bitcoind:29.2 not found: manifest unknown: manifest unknown
    ````
    
After facing these issues so far, I'd like to hand this PR back to you to make improvements. Please confirm that you have this feature fully functional on your end, then I can take another pass on it. Thanks again for what you have done so far.

kwsantiago commented:
Thank you for the feedback @jamaljsr . Addressed in 3246d68c4efa772edf643159cd45c4fe69a17b17

jamaljsr commented:
@kwsantiago I'm not sure how you are successfully running this on your machine. I'm still facing issues:

1. When trying to build the docker image:
```
$ docker build --build-arg BITCOIN_VERSION=29.2 --build-arg KNOTS_DATE=20251010 -t polarlightning/bitcoind-kn
ots:29.2 .
[+] Building 47.6s (9/12)                                                                                    docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                         0.0s
 => => transferring dockerfile: 1.21kB                                                                                       0.0s
 => [internal] load metadata for docker.io/library/debian:stable-slim                                                        1.1s
 => [auth] library/debian:pull token for registry-1.docker.io                                                                0.0s
 => [internal] load .dockerignore                                                                                            0.0s
 => => transferring context: 2B                                                                                              0.0s
 => [1/7] FROM docker.io/library/debian:stable-slim@sha256:7cb087f19bcc175b96fbe4c2aef42ed00733a659581a80f6ebccfd8fe3185a3d  0.0s
 => => resolve docker.io/library/debian:stable-slim@sha256:7cb087f19bcc175b96fbe4c2aef42ed00733a659581a80f6ebccfd8fe3185a3d  0.0s
 => => sha256:7cb087f19bcc175b96fbe4c2aef42ed00733a659581a80f6ebccfd8fe3185a3d 8.97kB / 8.97kB                               0.0s
 => => sha256:beecb6bc7f57ce868d4bc779278fff18f5695b2754044167ad3e0c141e2426da 1.04kB / 1.04kB                               0.0s
 => => sha256:898ab49d3e2bb1d49d32c0422a44593a94471289bcca1148e8ba8308cf803225 466B / 466B                                   0.0s
 => [internal] load build context                                                                                            0.0s
 => => transferring context: 1.08kB                                                                                          0.0s
 => [2/7] RUN apt-get update -y   && apt-get install -y curl gosu   && apt-get clean   && rm -rf /var/lib/apt/lists/* /tmp  43.4s
 => [3/7] RUN SYS_ARCH="$(uname -m)"   && MAJOR_VERSION="$(echo 29.2 | cut -d. -f1)"   && curl -SLO https://bitcoinknots.or  2.8s
 => ERROR [4/7] RUN curl -SLO https://raw.githubusercontent.com/bitcoin/bitcoin/master/contrib/completions/bash/bitcoin-cli  0.3s 
------                                                                                                                            
 > [4/7] RUN curl -SLO https://raw.githubusercontent.com/bitcoin/bitcoin/master/contrib/completions/bash/bitcoin-cli.bash   && mkdir /etc/bash_completion.d   && mv bitcoin-cli.bash /etc/bash_completion.d/bitcoin-cli.bash-completion   && curl -SLO https://raw.githubusercontent.com/scop/bash-completion/master/bash_completion   && mv bash_completion /usr/share/bash-completion/:             
0.108   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
0.108                                  Dload  Upload   Total   Spent    Left  Speed
100  4102  100  4102    0     0  33705      0 --:--:-- --:--:-- --:--:-- 33900
0.233 mkdir: cannot create directory '/etc/bash_completion.d': File exists
------
Dockerfile:19
--------------------
  18 |     
  19 | >>> RUN curl -SLO https://raw.githubusercontent.com/bitcoin/bitcoin/master/contrib/completions/bash/bitcoin-cli.bash \
  20 | >>>   && mkdir /etc/bash_completion.d \
  21 | >>>   && mv bitcoin-cli.bash /etc/bash_completion.d/bitcoin-cli.bash-completion \
  22 | >>>   && curl -SLO https://raw.githubusercontent.com/scop/bash-completion/master/bash_completion \
  23 | >>>   && mv bash_completion /usr/share/bash-completion/
  24 |     
--------------------
ERROR: failed to build: failed to solve: process "/bin/sh -c curl -SLO https://raw.githubusercontent.com/bitcoin/bitcoin/master/contrib/completions/bash/bitcoin-cli.bash   && mkdir /etc/bash_completion.d   && mv bitcoin-cli.bash /etc/bash_completion.d/bitcoin-cli.bash-completion   && curl -SLO https://raw.githubusercontent.com/scop/bash-completion/master/bash_completion   && mv bash_completion /usr/share/bash-completion/" did not complete successfully: exit code: 1

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/u2krscm2tun0zx45bi35mtfot
```

2. When running `yarn dev`:

<img width="2352" height="1232" alt="image" src="https://github.com/user-attachments/assets/ca66cf4b-04f5-425f-8f32-e2cd3d914e4c" />

Please resolve these errors, provide proper testing instructions and be sure that they work without any issues before requesting another round of review.

kwsantiago commented:
@jamaljsr both issues are resolved in 279137f8d6fba38c03bce38847c2a3c98e782529

Testing instructions below and in PR description.

### Build docker image
```
cd docker/bitcoind-knots
docker build --build-arg BITCOIN_VERSION=29.2 --build-arg KNOTS_DATE=20251110 -t
polarlightning/bitcoind-knots:29.2 .
```

### Run app
`yarn dev`

kwsantiago commented:
@jamaljsr any update on this? I left testing instructions on the last comment. Please let me know if anything else is needed.

Abdulkbk commented on src/utils/constants.ts:
```diff
@@ -56,10 +56,10 @@ export const BasePorts: Record<NodeImplementation, Record<string, number>> = {
-    p2p: 19444,
-    zmqBlock: 28334,
-    zmqTx: 29335,
+    rest: 18543,
```
you can squash this commit into the first one

Abdulkbk reviewed (changes requested):
Hey @kwsantiago, I looked into this feature, and I have a couple of initial feedback:

1. There are a couple of linter issues with some files, for example, `src/components/network/NetworkSetting.tsx` and `src/components/network/NewNetwork.tsx`.
2. Opening a terminal does not work currently.
3. I think the logo for knots should be different from that of bitcoind.
4. In some places, you use `bitcoind-knots`, while in others, you use `bitcoindKnots`, be consistent.
5. After creating a network, adding a knot using the drag-and-drop feature does not work.
6. Running `yarn ci` locally fails.

Generally, I think there's a bit more work needed to bring this to the finish line.

kwsantiago commented:
@Abdulkbk thank you for the feedback, it is greatly appreciated!

Addressed all feedback: fixed linter issues, terminal support, drag-and-drop, added distinct Knots logo, updated to v29.3, squashed commits, and yarn ci passes clean.

Note: Docker images need to be built and pushed to polarlightning/bitcoind-knots before end-to-end testing.

Abdulkbk commented:
> squashed commits

Oh no, this isn't what I was suggesting. Now I see only one commit, which makes it nearly impossible to review. You should break the commits, like you did earlier. My suggestion was to squash just the last commit into the first.

kwsantiago committed:
feat(docker): add bitcoind-knots Docker image

kwsantiago committed:
feat: add bitcoind-knots types, constants, and logo

kwsantiago committed:
feat: add bitcoind-knots network creation utilities

kwsantiago committed:
feat: integrate bitcoind-knots into store models and services

kwsantiago committed:
feat: add bitcoind-knots to UI components

Abdulkbk commented on src/utils/constants.ts:
```diff
@@ -230,6 +237,37 @@ export const dockerConfigs: Record<NodeImplementationWithSimln, DockerConfig> =
   },
+  'bitcoind-knots': {
+    name: 'Bitcoin Knots',
+    imageName: 'polarlightning/bitcoind-knots',
```
nit: should be `polarlightning/bitcoindknots` so that we can build with:
```bash
docker build --build-arg BITCOIN_VERSION=29.3 --build-arg KNOTS_DATE=20260210 -t polarlightning/bitcoindknots:29.3 .
```

which is also consistent with CLN.

Abdulkbk reviewed (changes requested):
Nice one. Tested the feature again. Have a few comments:

The terminal for knots still refuses to launch: 

<img width="1332" height="1035" alt="Image" src="https://github.com/user-attachments/assets/44bb2fef-538d-40d7-aba8-c5db55a66db5" />

Additionally, the unit tests do not cover the added code 100%, `yarn ci` coverage is at ~99, just a few lines untested.
<img width="759" height="53" alt="image" src="https://github.com/user-attachments/assets/84ed8d85-7d23-4220-954c-6d490fb2188c" />



Once I've finished testing the functionalities, I will examine the code more thoroughly.

kwsantiago committed:
fix: add bitcoind-knots to terminal node config

kwsantiago committed:
fix: correct bitcoind-knots image name and version numbers

Abdulkbk commented:
You should run `yarn ci` locally to ensure 100% coverage. There's still uncovered lines in `NewNetwork.tsx`, specifically line 63.

<img width="631" height="34" alt="image" src="https://github.com/user-attachments/assets/1e166a8f-82fb-4809-a282-42e8923fd7cc" />

Abdulkbk commented on src/utils/network.ts:
```diff
@@ -450,6 +489,7 @@ export const createNetwork = (config: {
   clightningNodes: number;
   eclairNodes: number;
   bitcoindNodes: number;
+  bitcoindKnotsNodes?: number;
```
To be consistent with all the other implementations, just use `bitcoindKnotsNodes: number` and then fix the lint issues by adding `bitcoindKnotsNodes: 0,` in the call sites that will complain.

Abdulkbk commented on src/utils/network.ts:
```diff
@@ -468,6 +508,7 @@ export const createNetwork = (config: {
     clightningNodes,
     eclairNodes,
     bitcoindNodes,
+    bitcoindKnotsNodes = 0,
```
then here will jiust be `bitcoindKnotsNodes,`

Abdulkbk commented on src/store/models/network.ts:
```diff
@@ -48,6 +49,7 @@ interface AddNetworkArgs {
   clightningNodes: number;
   eclairNodes: number;
   bitcoindNodes: number;
+  bitcoindKnotsNodes?: number;
```
here too, just make it `bitcoindKnotsNodes: number;`

Abdulkbk commented on src/store/models/network.ts:
```diff
@@ -324,6 +326,7 @@ const networkModel: NetworkModel = {
         clightningNodes: payload.clightningNodes,
         eclairNodes: payload.eclairNodes,
         bitcoindNodes: payload.bitcoindNodes,
+        bitcoindKnotsNodes: payload.bitcoindKnotsNodes || 0,
```
once you make it required, there won't be need for `|| 0`.

Abdulkbk commented on src/lib/docker/composeFile.ts:
You'll need to update `bitcoind(...)` to accept either `implementation` or `volumeDirName`, so that the volume field of the function uses the correct directory for the appropriate implementation. What happens now is that knots store its data in `networks/1/volumes/bitcoind/backend1/` instead of `networks/1/volumes/bitcoind-knots/backend1`.

Because of this, if you try to rename or remove a knot, the directory remains unchanged.

Abdulkbk commented on src/utils/constants.ts:
```diff
@@ -451,7 +451,7 @@ export const defaultRepoState: DockerRepoState = {
     'bitcoind-knots': {
       latest: '29.3',
-      versions: ['29.3', '29.2', '28.0', '27.0', '26.0'],
+      versions: ['29.3', '29.2', '28.1', '27.1', '26.1'],
```
You'll need to update nodes.json to0. It still has:
```javascript
  "bitcoind-knots": {
      "latest": "29.3",
      "versions": ["29.3", "29.2", "28.0", "27.0", "26.0"]
    },
```

Abdulkbk commented on src/utils/network.spec.ts:
```diff
@@ -511,7 +511,49 @@ describe('Network Utils', () => {
+      expect(network.nodes.bitcoin[0].version).toBe('29.3');
+    });
+
+    it('should adjust bitcoind-knots version for LND compatibility when target version is available', () => {
```
you can squash this commit into 172c3747fa7d08f9c8d5c4d9101fa6ae64df3a16

Abdulkbk commented on src/components/network/NewNetwork.spec.tsx:
```diff
@@ -124,6 +124,22 @@ describe('NewNetwork component', () => {
       expect(await findByText('asdf')).toBeInTheDocument();
     });
 
+    it('should show an error when lightning nodes exist without a bitcoin backend', async () => {
```
nit: can squash this into dfa78dbda318e7633b84c269dd3a97af20edbff2

Abdulkbk reviewed (changes requested):
Nice, this is getting closer to the finish line. I left some feedback, request for my review once you address them.

kwsantiago committed:
fix: address PR review feedback for bitcoind-knots support

- Make bitcoindKnotsNodes required (not optional) in interfaces
- Fix volume directory so knots uses bitcoind-knots/ not bitcoind/
- Update nodes.json versions to match constants.ts
- Add bitcoindKnotsNodes: 0 to all call sites

Abdulkbk commented on docker/README.md:
```diff
@@ -47,6 +47,24 @@ $ docker buildx build --platform linux/amd64,linux/arm64 --build-arg BITCOIN_VER
+### Tags
+
+- `29.2` ([bitcoind-knots/Dockerfile](https://github.com/jamaljsr/polar/blob/master/docker/bitcoind-knots/Dockerfile))
+- `28.0` ([bitcoind-knots/Dockerfile](https://github.com/jamaljsr/polar/blob/master/docker/bitcoind-knots/Dockerfile))
```
The Readme still shows 29.2, 28.0, 27.0, 26.0, while nodes.json and constant.ts have 29.3, 29.2, 28.1, 27.1, 26.1.

Abdulkbk commented on src/shared/types.ts:
```diff
@@ -82,6 +82,16 @@ export interface BitcoindNode extends BitcoinNode {
   };
 }
 
+export interface BitcoindKnotsNode extends BitcoinNode {
```
This is not used anywhere.

Abdulkbk reviewed (changes requested):
Just a couple of feedback, otherwise LGTM.

kwsantiago committed:
fix: update README versions and remove unused BitcoindKnotsNode interface

Abdulkbk reviewed (approved):
LGTM ⚡️and thanks for working on this feature.

Tested across macOS, Windows, and Linux.

cc: @jamaljsr for additional review.

Abdulkbk commented:
Hey @kwsantiago. I had a chat with @jamaljsr regarding this PR, and he pointed out an issue that he had explained in bullet 2 of this [comment](https://github.com/jamaljsr/polar/pull/1303#issuecomment-3505998264).

Basically, if you try to add `LND 0.18.3-beta` to `Knots 27.1`, you'd get a `backendCompatError`, even though `LND 0.18.3-beta` supports `Bitcoin core 27.0`. It also seems true for `Bitcoin core`, but we didn't experience this issue because we didn't have point releases for it.

Maybe we should update our compat check to use major releases so that `29.0` works for all `29.x` release series.

kwsantiago committed:
fix: use major version for backend compat check

kwsantiago commented:
@Abdulkbk Good catch, thanks for the heads up. Updated filterCompatibleBackends in 870985f to compare by major version, so Knots 27.1 now resolves against LND's 27.0 target (and likewise for any future Core point releases).

Abdulkbk reviewed (approved):
re-tACK 870985fed82e035ce3559e81cdfd1c1d331ffb3d
